core/src/main/java/io/grpc/Context.java

```diff
@@ -566,65 +585,53 @@ private Object lookup(Key<?> key) {
   public static final class CancellableContext extends Context {
 
     private boolean cancelled;
-    private Throwable cause;
-    private final Context dummy;
-    private ScheduledFuture<?> scheduledFuture;
+    private Throwable cancellationCause;
+    private final Context uncancellableSurrogate;
+    private ScheduledFuture<?> pendingDeadline;
 
     /**
      * Create a cancellable context that does not have a deadline.
      */
     private CancellableContext(Context parent) {
       super(parent, EMPTY_ENTRIES, true);
-      // Create a dummy that inherits from this to attach so that you cannot retrieve a
+      // Create a surrogate that inherits from this to attach so that you cannot retrieve a
       // cancellable context from Context.current()
-      dummy = new Context(this, EMPTY_ENTRIES);
+      uncancellableSurrogate = new Context(this, EMPTY_ENTRIES);
     }
 
     /**
      * Create a cancellable context that has a deadline.
      */
-    private CancellableContext(Context parent, long delayNanos,
+    private CancellableContext(Context parent, Deadline deadline,
         ScheduledExecutorService scheduler) {
       this(parent);
-      scheduledFuture = scheduler.schedule(new Runnable() {
-        @Override
-        public void run() {
-          cancel(new TimeoutException("context timed out"));
-        }
-      }, delayNanos, TimeUnit.NANOSECONDS);
+      Deadline parentDeadline = DEADLINE_KEY.get(parent);
+      if (parentDeadline == null || deadline == deadline.minimum(parentDeadline)) {
```

Comment:
I'm not sure how much you want to deal with these cases. Note that if `deadline == parentDeadline` this schedules a timer.